### PR TITLE
Add producer to StableHLO portable artifacts.

### PR DESCRIPTION
**Pull request by GleasonK**
Producers reads `StableHLO_v<target_version>`, for example:

```
$ stablehlo-translate --serialize --target=0.9.0 file.mlir
ML?RStableHLO_v0.9.0[...]
```

Also moved some logic around and added `minimum` as supported target version now that we have a getMinimumVersion API.

Currently there are no APIs that inspect the producer string, so this is purely debug info and does not impact forward or backward compatibility.

### CODE REVIEW

File: stablehlo/tests/stablehlo_legalize_to_vhlo.0_9_0.mlir [hunk 1]
```diff
@@ -1,6 +1,7 @@
 // RUN: stablehlo-opt --mlir-print-op-generic %s.bc | FileCheck %s
 // RUN: stablehlo-translate --deserialize %s.bc | stablehlo-translate --serialize --target=0.9.0 | stablehlo-opt --mlir-print-op-generic | FileCheck %s
 // RUN: diff <(stablehlo-translate --deserialize %s.bc | stablehlo-opt) <(stablehlo-opt %s)
+// RUN: diff <(stablehlo-translate --serialize --target=minimum %s | stablehlo-translate --deserialize | stablehlo-opt) <(stablehlo-opt %s)
```
Comment (burmako) [inline stablehlo/tests/stablehlo_legalize_to_vhlo.0_9_0.mlir:4]: Am I right in thinking that the purpose of this RUN directive is orthogonal to this specific PR?

File: stablehlo/transforms/VhloToVersion.cpp [hunk 2]
```diff
@@ -67,15 +67,10 @@ class VhloToVersionConverter : public TypeConverter {
   }
 };
 
-FailureOr<Version> parseTargetVersion(llvm::StringRef versionRef) {
-  if (versionRef == "current") return Version::getCurrentVersion();
-  return Version::fromString(versionRef);
-}
```
Comment (burmako) [inline stablehlo/transforms/VhloToVersion.cpp:73]: If we're cleaning up version alias parsing, we could also clean up similar logic in getMinVersion / getMaxVersion methods defined in TableGen for VHLO program elements.

File: stablehlo/tests/stablehlo_legalize_to_vhlo.0_9_0.mlir (hunk 1 above)
Comment (GleasonK) [reply stablehlo/tests/stablehlo_legalize_to_vhlo.0_9_0.mlir:4]: Test that "minimum" as a serialization target works? Seems related to this PR?

File: stablehlo/api/PortableApi.cpp [hunk 3]
```diff
@@ -62,15 +62,12 @@ LogicalResult deserializePortableArtifact(StringRef artifactStr,
...
-  // moment of writing. The forward compatibility provided by it might be
-  // ultimately unneeded, but until we've established that, let's play it safe.
-  BytecodeWriterConfig writerConfig;
-  writerConfig.setDesiredBytecodeVersion(1);
-  return writeBytecodeToFile(*module, os, writerConfig);
+  // This bytecode does not need to specify verison number or header.
+  // This function, like serializePortableArtifact using module string,
+  // provides the reliability of passing / receiving textual assembly
+  // format. Bytecode is used here since it is more compact and faster
+  // to read / write in APIs that require passing strings.
```
[the remark this answers is not in the document]
Comment (GleasonK) [reply stablehlo/api/PortableApi.cpp]: Done.

File: stablehlo/dialect/Version.cpp [hunk 4]
```diff
@@ -49,6 +49,8 @@ static FailureOr<std::array<int64_t, 3>> extractVersionNumbers(
 }  // namespace
 
 FailureOr<Version> Version::fromString(llvm::StringRef versionRef) {
+  if (versionRef == "current") return Version::getCurrentVersion();
+  if (versionRef == "minimum") return Version::getMinimumVersion();
```
[the remark this answers is not in the document]
Comment (GleasonK) [reply stablehlo/dialect/Version.cpp:53]: Agree. But making as a part of this change to avoid the temporary redundancy.

There is some oddity that this aliasing is now needed in two places:
1. Calls to serialize via `stablehlo-translate --serialize --target=current`
2. Explicit pass option to `--vhlo-to-version='target=current'`

Making this change now avoids the redundancy.

File: stablehlo/dialect/Serialization.cpp [hunk 5]
```diff
@@ -31,6 +31,9 @@ LogicalResult serializePortableArtifact(ModuleOp module,
                                         StringRef targetVersion,
                                         raw_ostream& os) {
   MLIRContext* context = module.getContext();
+  auto version = vhlo::Version::fromString(targetVersion);
+  if (failed(version))
```
[the remark this answers is not in the document]
Comment (GleasonK) [reply stablehlo/dialect/Serialization.cpp]: This is used later to write the producer string. Renamed to `targetVersionStr` and `targetVersion`.

File: stablehlo/dialect/Version.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (GleasonK) [reply stablehlo/dialect/Version.cpp]: Actually.. we don't need to support aliases in VhloToVersion.. Can make sure call sites use explicit version numbers and bump this logic into `Serialization.cpp`.

[the remark this answers is not in the document]
Comment (GleasonK) [reply stablehlo/dialect/Version.cpp]: Moved this logic to `Serialization.cpp` and only accept fully qualified version numbers in `VhloToVersion`.

File: stablehlo/dialect/Serialization.cpp [hunk 6]
```diff
@@ -26,11 +26,23 @@ limitations under the License.
...
+FailureOr<vhlo::Version> parseTargetVersion(StringRef targetVersion) {
+  if (targetVersion == "current") return vhlo::Version::getCurrentVersion();
+  if (targetVersion == "minimum") return vhlo::Version::getMinimumVersion();
+  return vhlo::Version::fromString(targetVersion);
+}
+}  // namespace
 
 LogicalResult serializePortableArtifact(ModuleOp module,
-                                        StringRef targetVersion,
+                                        StringRef targetVersionStr,
```
Comment (burmako) [inline stablehlo/dialect/Serialization.cpp:38]: How about renaming the local variable instead, to something like `parsedTargetVersion`? Otherwise, for consistency reasons we'd have to rename the parameter in Serialization.h, as well as in PortableApi.h, as well as the python API (which would also involve bumping the API version).

upd: Given the conversation that we just had, it sounds like we won't need the local variable anymore, so I guess this comment is moot now.

File: stablehlo/dialect/Version.h [hunk 7]
```diff
@@ -35,6 +35,8 @@ class Version {
  public:
   /// Convenience method to extract major, minor, patch and create a Version
   /// from a StringRef of the form `#.#.#`. Returns failure if invalid string.
+  /// Returns current version if input string is "current", and minimum version
+  /// if input string is "minimum".
```
Comment (burmako) [inline stablehlo/dialect/Version.h:39]: This comments seems stale.

File: stablehlo/tests/stablehlo_legalize_to_vhlo.0_9_0.mlir (hunk 1 above)
Comment (burmako) [reply stablehlo/tests/stablehlo_legalize_to_vhlo.0_9_0.mlir:4]: As discussed just now, let's delete this RUN directive?

File: stablehlo/dialect/Serialization.cpp [hunk 8]
```diff
@@ -26,11 +26,23 @@ limitations under the License.
 
 namespace mlir {
 namespace stablehlo {
+namespace {
+FailureOr<vhlo::Version> parseTargetVersion(StringRef targetVersion) {
+  if (targetVersion == "current") return vhlo::Version::getCurrentVersion();
+  if (targetVersion == "minimum") return vhlo::Version::getMinimumVersion();
+  return vhlo::Version::fromString(targetVersion);
+}
+}  // namespace
```
Comment (burmako) [inline stablehlo/dialect/Serialization.cpp:35]: Let's move this logic into stablehlo-translate?

**Review by burmako — CHANGES_REQUESTED** [review]
